presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.matching;
+
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Streams;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class MatchablesStore<T extends Matchable>
```

Comment:
I do not think we should use `<T extends Matchable>` here. It unnecessary adds class hierarchy dependency for users of `MatchableStore`. 

It should be 
```
public class MatchablesStore<T>
```

And in `Builder` we could have two methods
```
register(T matchable, Pattern pattern) {
   ....
} 
register(T matchable) {
   checkArgument(matchable instanceOf Matchable);
   register(matchable, ((Matchable)matchable).getPattern())
}